examples/cjs.py

```diff
@@ -0,0 +1,288 @@
+        return (first_capture_mask, z), None
+
+    z = jnp.ones(N, dtype=jnp.int32)
+    # we use this mask to eliminate extraneous log probabilities
+    # that arise for a given individual before its first capture.
+    first_capture_mask = capture_history[:, 0].astype(bool)
+    # NB swapaxes: we move time dimension of `capture_history` to the front to scan over it
+    scan(transition_fn, (first_capture_mask, z), jnp.swapaxes(capture_history[:, 1:], 0, 1))
+
+
+# %%
+# In our second model variant there is a time-varying survival probability phi_t for
+# T-1 of the T time periods of the capture data; each phi_t is treated as a fixed effect.
+
+
+def model_2(capture_history, sex):
+    N, T = capture_history.shape
+    rho = numpyro.sample("rho", dist.Uniform(0.0, 1.0))  # recapture probability
+
+    def transition_fn(carry, y):
+        first_capture_mask, z = carry
+        # note that phi_t needs to be outside the plate, since
+        # phi_t is shared across all N individuals
+        phi_t = numpyro.sample("phi", dist.Uniform(0.0, 1.0))
+
+        with numpyro.plate("animals", N, dim=-1):
+            with handlers.mask(mask=first_capture_mask):
+                mu_z_t = first_capture_mask * phi_t * z + (1 - first_capture_mask)
+                # NumPyro exactly sums out the discrete states z_t.
+                z = numpyro.sample("z", dist.Bernoulli(dist.util.clamp_probs(mu_z_t)))
+                mu_y_t = rho * z
+                numpyro.sample("y", dist.Bernoulli(dist.util.clamp_probs(mu_y_t)), obs=y)
+
+        first_capture_mask = first_capture_mask | y.astype(bool)
+        return (first_capture_mask, z), None
+
+    z = jnp.ones(N, dtype=jnp.int32)
+    # we use this mask to eliminate extraneous log probabilities
+    # that arise for a given individual before its first capture.
+    first_capture_mask = capture_history[:, 0].astype(bool)
+    # NB swapaxes: we move time dimension of `capture_history` to the front to scan over it
+    scan(transition_fn, (first_capture_mask, z), jnp.swapaxes(capture_history[:, 1:], 0, 1))
+
+
+# %%
+# In our third model variant there is a survival probability phi_t for T-1
+# of the T time periods of the capture data (just like in model_2), but here
+# each phi_t is treated as a random effect.
+
+
+def model_3(capture_history, sex):
+    N, T = capture_history.shape
+    phi_mean = numpyro.sample("phi_mean", dist.Uniform(0.0, 1.0))  # mean survival probability
+    phi_logit_mean = logit(phi_mean)
+    # controls temporal variability of survival probability
+    phi_sigma = numpyro.sample("phi_sigma", dist.Uniform(0.0, 10.0))
+    rho = numpyro.sample("rho", dist.Uniform(0.0, 1.0))  # recapture probability
+
+    def transition_fn(carry, y):
+        first_capture_mask, z = carry
+        with handlers.reparam(config={"phi_logit": LocScaleReparam(0)}):
+            phi_logit_t = numpyro.sample("phi_logit", dist.Normal(phi_logit_mean, phi_sigma))
```

Comment:
@martinjankowiak I think the only change in models is the use of reparam here. Without this, it is pretty slow because each trajectory took 1024 leapfrog steps... Other changes are equivalent to Pyro:
+ Pyro clamps `probs` by default. Without clamping, things like `Bernoulli(0).log_prob(1)` will lead to finite potential energy but `nan` gradients.
+ It is tricky to use `if else` logic in JAX, so I unrolled the first step.
+ for loop -> transition_fn + scan